### ISSUE TITLE
Python is generating very bad IRs

### ISSUE THREAD

**Issue by tpoterba** [hail-is/hail#4001]
consider:
```ht.annotate(x=5)```

Here's the IR:

```
(TableMapRows (idx) 1
  (TableRange 5 5)
  (Let __uid_1
    (InsertFields
      (SelectFields ()
        (Ref Struct{idx:Int32} row))
      (x
        (I32 5)))
    (InsertFields
      (MakeStruct
        (idx
          (GetField idx
            (Ref Struct{idx:Int32} row))))
      (x
        (GetField x
          (Ref Struct{x:Int32} __uid_1))))))
```

Cotton added some optimizer rules, but they're not always sufficient. We need to fix this

**Comment by tpoterba** [comment]
the problem is related to the current level of abstraction between annotate+select, as well as the handling of key fields.

**Comment by tpoterba** [comment]
Currently this is optimized to:

```
(TableMapRows (idx) 1
  (TableRange 5 5)
  (MakeStruct
    (idx
      (GetField idx
        (Ref Struct{idx:Int32} row)))
    (x
      (I32 5))))
```

But clearly it should be an InsertFields, not a MakeStruct.

**Comment by catoverdrive** [comment]
This specific example is fixed by #4042.
